### ISSUE TITLE
Typing almost non-existent

### ISSUE THREAD

**Issue by brianreinhold** [micropython/micropython-lib#791]
Is there a place to open up a discussion versus an issue?

For someone using this library for the first time it makes it extremely difficult when a vast majority of variables are typed as Incomplete or None. I have to admit that I dislike dynamic typing as it is hard to follow someone else's code when nothing is typed. The same goes for the examples.

I am trying to make a decision as whether to use the low-level bluetooth or this package on the PICO-W. I get the impression that the low-level library is better typed and it follows the Bluetooth spec more closely. Since I understand the latter quite well and have done a lot of work in C and Java in BTLE in various platforms the low-level option may be a better approach and easier to follow. But it comes with a warning that it is subject to change, so I am not really sure where to go.

is there any active discussion forum for this library and documentation that describes the API and variables (besides going to the source code)? The ReadMe leaves out a lot of necessary details. I will need to write a central/client for new BTLE profiles (GHS in particular) and a host of proprietary devices. Are there things that one cannot do at this level that one can do using the low-level?

Thanks

**Comment by mattytrentini** [comment]
There is a [discussion board](https://github.com/orgs/micropython/discussions/) on the main micropython repository. I can't convert this directly into a discussion post but I could transfer it to a micropython issue (different repo) and then to a discussion. Let me know if you'd prefer that!

You should also be aware of the [aioble library](https://github.com/micropython/micropython-lib/tree/master/micropython/bluetooth/aioble) if you're not already - it's a much more productive way to build BLE clients. 

As for typing, it wouldn't hurt to add typing information to the low level BLE library but that may be best done with stubs. I don't think it's possible to add typing information to MicroPython features implemented solely in C. Perhaps [micropython-stubber](https://github.com/Josverl/micropython-stubber/) can help?

**Comment by brianreinhold** [comment]
@mattytrentini thanks for your reply. I will check out these boards. I should point out that I need to build a generic health device client which supports not only BT SIG health device standards including the new GHS but proprietary devices that can behave in a who-knows-what way. Its the latter that poses the big challenge!  If the high level API allows me to do anything I could do in the low level API then it probably makes more sense to use the high level API even though I can see the correspondence between the low-level API and BTLE GAP/GATT functionality much more clearly. I have done a lot of work on the Nordic and Renesas platforms (C) and Android (Java) and these APIs look quite similar to the low-level API with its basic callback structure so that familiarity biases me toward the low-level choice.

**Comment by jonnor** [comment]
https://github.com/Josverl/micropython-stubs provides typing information for huge parts of MicroPython modules. I am pretty sure it inlcudes aioble.

Proposing to close this issue. As mentioned, the discussions is better place for more general/open ended. Feel free to open one there. And also to share what you have learned :)

**Comment by brianreinhold** [comment]
If I had known about that at the time I would have surely taken it up. Our company needed to have an aioble capable of supporting a client application on the new micropython Rasberry PI PICO for bluetooth low energy health devices, and that means the ability to pair and bond. Did a lot of work to add support for pairing and bonding (I think it was Bluekitchen's bt stack that needed to be integrated into the mix). Added typing on the fly, needed a few more APIs, but eventually got it to work. We did not commit our back (as that in itself would be a time consuming effort that the company could not afford to spend resources on) but also because we only handled the client side.

But that was a few years ago now and the company has shut down and personally, in my retirement years, I have returned to research meteorology. I am sure our work was done on an outdated version of MP and the Bt Stack by now and is probably of no use (though I still have it).

**Comment by Josverl** [comment]
@brianreinhold,

If you have something that might be useful, I'd appreciate a link or a blob in an issue or discussion at the MicroPython-stubs repo for me to sort out. Even uncommon examples or code snippets are valuable to me as they allow me to validate, and often improve, the stubs. 

If not, then thanks for your earlier contributions.

**Comment by brianreinhold** [comment]
Josverl,

I don't  know if this will be useful, but I attach a few repos. One is our PHG (Personal Health Gateway) implementation on the PICO. I had to make radical changes to support pairing in aioble so what I ended up doing is copying the entire aioble library into our own project and I made the changes there.  Some earlier attempts also remain in the original but in the end we never used the MP aioble in our PCIO PHG project. 

Also needed to modify the bt stack interface and btStack C code to support pairing.

BLE Health devices can bond and often do a bonded reconnect and many devices invoke the security request (both on a first time connect and a bonded reconnect). Others wait for the gateway client to initiate bonding (I know of no device that pairs without bonding) and encryption on a bonded reconnect. One also needs to handle the ble error 5 (insufficient authentication) and initiate bonding. Unfortunately these can happen at anytime which makes it a real pain. When working on the Generic Health Device Standard one thing I got into the standard was to get around this unpredictable events from the health device during bonding or bonded reconnects.

I tried to put notes in the micropython C code and btstack code where changes were made. Look for //Brian  (not the only Brian in there as it turns out). 

Looks like I can't upload the files (as zips). Must be too big. I wanted to keep the git bundle so you could see the history and changes. Maybe there is another way?

**Comment by Josverl** [comment]
@brianreinhold 

EDIT:

I was still typing - then we were kicked out - I must have set the time wrong.
I have been able to download the repo - and unzipped it.
I can see the full history so i can go spelunking.

Thank you very much sir !
